web/server/models.py

```diff
@@ -362,3 +372,136 @@ class ArtifactChange(BaseModel):
     path: str
     type: t.Optional[ArtifactType] = None
     file: t.Optional[File] = None
+
+
+class ReportBackfillProgress(BaseModel):
+    completed: int
+    total: int
+    view_name: str
+    start: TimeLike
+    end: t.Optional[TimeLike] = None
+
+
+class ReportTestsResult(BaseModel):
+    message: str
+
+
+class ReportTestDetails(ReportTestsResult):
+    details: str
+
+
+class ReportTestsFailure(ReportTestsResult):
+    total: int
+    failures: int
+    errors: int
+    successful: int
+    dialect: str
+    details: t.List[ReportTestDetails]
+    traceback: str
+
+
+class ReportMeta(BaseModel):
+    status: ReportStatus = ReportStatus.init
+    start_at: int = Field(default_factory=now_timestamp)
+    stop_at: t.Optional[int] = None
+    done: bool = False
+
+    @property
+    def duration(self) -> int | None:
+        return self.stop_at - self.start_at if self.start_at and self.stop_at else None
+
+    def dict(self, *args: t.Any, **kwargs: t.Any) -> t.Dict[str, t.Any]:
+        data = super().dict(*args, **kwargs)
+        data["duration"] = self.duration
+        return data
+
+
+class Stage(BaseModel):
+    meta: ReportMeta = Field(default_factory=ReportMeta)
+
+    def stop(self, success: bool = True) -> None:
+        if success:
+            self.meta.status = ReportStatus.success
+        else:
+            self.meta.status = ReportStatus.fail
+
+        self.meta.stop_at = now_timestamp()
+        if self.meta.start_at and self.meta.stop_at:
+            self.meta.done = True
+
+
+class BackfillDetails(BaseModel):
+    model_name: t.Optional[str] = None
+    view_name: str
+    interval: t.Tuple[str, str]
+    batches: int
+
+
+class BackfillTask(BaseModel):
+    completed: int
+    total: int
+    view_name: str
+    start: TimeLike
+    end: t.Optional[TimeLike] = None
+
+
+class ReportStage(Stage):
+    def update(self, data: t.Dict[str, t.Any]) -> None:
+        for k, v in data.items():
+            setattr(self, k, v)
+
+
+class ReportStagePlanValidation(ReportStage):
+    start: t.Optional[TimeLike] = None
+    end: t.Optional[TimeLike] = None
+
+
+class ReportStagePlanChanges(ReportStage):
+    added: t.Optional[t.Set[str]] = None
+    removed: t.Optional[t.Set[str]] = None
+    modified: t.Optional[ModelsDiff] = None
+
+
+class ReportStagePlanBackfills(ReportStage):
+    models: t.Optional[t.List[BackfillDetails]] = None
+
+
+class ReportStagePlanApplyCreation(ReportStage):
+    total_tasks: int
+    num_tasks: int
+
+
+class ReportStagePlanApplyRestate(ReportStage):
+    pass
+
+
+class ReportStagePlanApplyBackfill(ReportStage):
+    queue: t.Set[str] = set()
+    tasks: t.Dict[str, BackfillTask] = {}
+
+
+class ReportStagePlanApplyPromote(ReportStage):
+    total_tasks: int
+    num_tasks: int
+    target_environment: str
+
+
+class Report(Stage):
+    environment: str
+    options: t.Optional[t.Dict[str, t.Any]] = None
```

Comment:
Is it possible to enumerate all the possible options in something like a ReportOptions Pydantic model?